core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1383,6 +1383,8 @@ private[spark] class DAGScheduler(
                 if (!job.finished(rt.outputId)) {
                   job.finished(rt.outputId) = true
                   job.numFinished += 1
+                  taskScheduler.markPartitionIdAsCompletedAndKillCorrespondingTaskAttempts(
```

Comment:
Does this require the newest `TaskSet` has been created in `TaskSetScheduler` ? Can we make sure of that ?

Reply:
No, this method does not require that the newest TaskSet be created.

Reply:
If not, how does it know about the completed partition ? Or say, is it possible this happens druing the time we have put the stage into the resubmit queue but haven't submited to TaskScheduler(so there's no newest TaskSet) ?

Reply:
Ahh I see what you are saying. Yes, this method assumes that the newest TaskSet has been created. Yes, indeed a tiny possibility exists that when this method gets called, possibly the new TaskSet might have not been added to `taskSetsByStageIdAndAttempt`. However, when I wrote the code, my assumption was that there is always a small delay for the task completion event to propagate to the DAGScheduler. I have tested this code by reproducing FetchFailures multiple times and when this method is called, the new TaskSet is always present so there has not yet been an instance when the race condition described above has occurred, while before this fix, I was able to reproduce the bug like 4 out of 5 times. Still, a valid point.

Reply:
Yeah, I agree this could be a extremely rare case and your work has proved that this pr is really a effective fix.

And as I mentioned above, if we could record this completed partition in `TaskScheduler` at this point and telling others TaskSets  about the completed patitions once they're created in `TaskScheduler`, we may avoid the potential issue totally. But, I think we don't need to do this right now since your fix has already proved to be effective. Instead, how about leaving some comments to explain the potential issue ? Thus, we may easily fix it once we really hit it.

Reply:
Makes sense, have updated the comment. Thank you.